Adding LOBPCG solver in svds in addition to ARPACK

lobpcg opened the pull request:
<!-- 
Adding LOBPCG solver in svds in addition to ARPACK
-->

#### Reference issue
Fixes #9275 

#### What does this implement/fix?
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html is a naive implementation using ARPACK as an eigensolver on A.H * A or A * A.H, depending on which one is more efficient.

It is simple and beneficial to add an alternative to ARPACK solver https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lobpcg.html again just on A.H * A or A * A.H, depending on which one is more efficient. It is demonstrated in MATLAB that LOBPCG may outperform ARPACK for such a use, see comments at https://www.mathworks.com/matlabcentral/fileexchange/48-lobpcg-m

As a free bonus:
- the new rmatmat option has been added here, after being merged in #10849 (comment) 
- resolve https://github.com/scipy/scipy/issues/9643 
- supersede the stuck PR https://github.com/scipy/scipy/pull/8169/

lobpcg commented:
It is done - could someone please review? 
I have also beautified the unit test code a bit, unrelated to this PR, just to make pylint happier.

mdhaber commented:
I don't know the algorithm - is there any chance this will work better than ARPACK at finding the smallest singular values when there are multiple nearly-zero singular values?
I don't have time to review in detail ATM, but I might be able to take a look this weekend.

lobpcg commented:
> I don't know the algorithm - is there any chance this will work better than ARPACK at finding the smallest singular values when there are multiple nearly-zero singular values?
> I don't have time to review in detail ATM, but I might be able to take a look this weekend.

Possibly, you can just try running it. To predict the outcome a priori , I would need to know how the small eigenvalues are distributed and how many there are.  Thanks!

lobpcg commented:
It is done - could someone please review?

ilayn commented on scipy/sparse/linalg/eigen/arpack/arpack.py:
```diff
@@ -1822,8 +1827,26 @@ def matvec_XH_X(x):
+        else:
+            X = np.random.RandomState(52).randn(min(A.shape), k)
+
+        eigvals, eigvec = lobpcg(XH_X, X, tol=tol ** 2, maxiter=maxiter,
```
Is there any reason why we are squaring the tolerances? This should be at least documented, (personal opinion, shouldn't be done at all).

lobpcg replied:
I just kept it the same as in eigsh below...

lobpcg replied:
Please advise?

lobpcg replied:
No change, mark as resolved - please see similar one below

ilayn commented on scipy/sparse/linalg/eigen/arpack/arpack.py:
```diff
@@ -1822,8 +1827,26 @@ def matvec_XH_X(x):
+                          largest=largest)
+
+    else:
+        eigvals, eigvec = eigsh(XH_X, k=k, tol=tol ** 2, maxiter=maxiter,
```
see above for tolerance squaring

lobpcg replied:
I do not know if it is documented. I suppose that the reason might be that one measures the norm of the residual using the normal matrix A'*A. The serious issue is if the tolerance gets so small that the solver could never reach it.  I have not looked at this. It may be a good issue to investigate and possibly file, but I hope unrelated to this PR, since the unit tests pass...

lobpcg replied:
Please advise?

pv replied:
I'd consider this issue separately.

lobpcg replied:
Thanks! No change, mark as resolved

ilayn commented on scipy/sparse/linalg/eigen/arpack/tests/test_arpack.py:
```diff
@@ -136,26 +136,26 @@ def assert_allclose_cc(actual, desired, **kw):
             if OPpart is None:
-                reval = 1. / (eval - sigma)
+                reval = 1. / (eigenvalues - sigma)
             elif OPpart == 'r':
```
Is there any chance `OPpart` is none of these options to catch?

lobpcg replied:
I agree, but this is actually unrelated to the PR. I only renamed "eval" into "eigenvalues" to make the pylint happier.  Also, this file is only a test, where I am sure there are issues like this one in many-many spots...

lobpcg replied:
Please advise?

pv replied:
I think no change is needed here (the test would just run into NameError)

lobpcg replied:
Thanks! Mark as resolved

lobpcg commented:
@ilayn -thanks for your review! Please see my replies.

lobpcg commented:
> 
> 
> No detailed review right now, and I haven't read the other comments, but two observations based on testing:
> 
>     * for `which='LM'`, eigenvalues are being returned in descending order, whereas arpack returned them ascending order
> 
>     * I'm getting an error when I don't specify `maxiter`
> 
> 
> This is based on testing with the matrix `bcspwr05` from the Singular Matrix Database.
> (Unfortunately, this does not fare much better than ARPACK on this matrix for `which='SM'`. For `k=1`, ARPACK errors out, but LOBPCG reports an incorrect value.)
> I'll try to provide a minimal example, but thought I'd report the issues in the meantime.

@mdhaber Great, thanks for testing! 
I'll check  which='LM' - makes me also curious how come the unit tests pass. 

Could you please be a bit more specific about  error when don't specify maxiter?

You do not want to run LOBPCG with the defaults to get accurate answers for difficult problems...

mdhaber commented:
```
import scipy.sparse
import numpy as np
from scipy.sparse.linalg import svds

np.random.seed(0)
A = scipy.sparse.diags(np.random.rand(1000), 0)
svds(A, solver='lobpcg')
```
Output:
```
Traceback (most recent call last):
  File "test2.py", line 15, in <module>
    svds(A, solver='lobpcg')
  File "/home/scipy/scipy/sparse/linalg/eigen/arpack/arpack.py", line 1845, in svds
    largest=largest)
  File "/home/scipy/scipy/sparse/linalg/eigen/lobpcg/lobpcg.py", line 417, in lobpcg
    while iterationNumber < maxIterations:
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

lobpcg commented:
> 
> 
> ```
> import scipy.sparse
> import numpy as np
> from scipy.sparse.linalg import svds
> 
> np.random.seed(0)
> A = scipy.sparse.diags(np.random.rand(1000), 0)
> svds(A, solver='lobpcg')
> ```
> 
> Output:
> 
> ```
> Traceback (most recent call last):
>   File "test2.py", line 15, in <module>
>     svds(A, solver='lobpcg')
>   File "/home/scipy/scipy/sparse/linalg/eigen/arpack/arpack.py", line 1845, in svds
>     largest=largest)
>   File "/home/scipy/scipy/sparse/linalg/eigen/lobpcg/lobpcg.py", line 417, in lobpcg
>     while iterationNumber < maxIterations:
> TypeError: '<' not supported between instances of 'int' and 'NoneType'
> ```

Great, many thanks! Appears to be a good bug in lobpcg.py I'll fix it.

mdhaber commented:
Regarding the order, the [svds](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) documentation doesn't actually guarantee an order for the singular values, but ARPACK seems to be consistent. If there is a reason for this, it would be nice if LOBPCG followed suit.

Another observation: with `solver='blarg'`, `svds` defaults to using ARPACK with no message. For functions that accept a `method` argument, there is some precedent for erroring out when the method is invalid (e.g. `minimize`, `linprog`). There may be conflicting precedent, but just something to consider.

lobpcg commented:
> 
> 
> Regarding the order, the [svds](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) documentation doesn't actually guarantee an order for the singular values, but ARPACK seems to be consistent. If there is a reason for this, it would be nice if LOBPCG followed suit.
> 
> Another observation: with `solver='blarg'`, `svds` defaults to using ARPACK with no message. I discovered this when I misspelled `lbpcg`). For functions that accept a `method` argument, there is some precedent for erroring out when the method is invalid (e.g. `minimize`, `linprog`). There may be conflicting precedent, but just something to consider.

@mdhaber thanks for the review! Could you please have another look? I think that I have fixed both issues.

mdhaber commented on scipy/sparse/linalg/eigen/arpack/arpack.py:
```diff
@@ -1822,8 +1827,29 @@ def matvec_XH_X(x):
+        eigvals, eigvec = lobpcg(XH_X, X, tol=tol ** 2, maxiter=maxiter,
+                                 largest=largest)
+
+    elif solver == 'arpack' or solver == None:
```
```suggestion
    elif solver == 'arpack' or solver is None:
```

lobpcg replied:
Thanks very much! Done

lobpcg commented:
Please advise if there is anything else I should do.

mdhaber commented:
I'm not able to review in detail for a while.

lobpcg commented:
> I'm not able to review in detail for a while.

@mdhaber thanks again for your review! I made all the changes that you requested in this PR. Could you please consider thus removing your block "Changes requested"? May be other experts would review.

mdhaber commented:
>  Could you please consider thus removing your block "Changes requested"? 

I will when I can just verify that the changes were made. Unit tests verifying the changes would be nice, if you think they're appropriate. I don't think it's a bad idea to include them, since we don't want bugs to come back.

lobpcg commented:
> 
> 
> > Could you please consider thus removing your block "Changes requested"?
> 
> I will when I can just verify that the changes were made. Unit tests verifying the changes would be nice, if you think they're appropriate. I don't think it's a bad idea to include them, since we don't want bugs to come back.

One change is simply explicit sorting in a few lines of code just before the return. Sorting is an undocumented feature anyway, so probably should not come with a unit test. 

The other change is in a single line copy/ pasted from your own suggestion.

lobpcg commented:
rmatmat option has been added in https://github.com/scipy/scipy/pull/10849#issuecomment-538102582 so I'll try to see if I can use it here, while waiting for the reviews. 
I cannot compile and test scipy locally, so have to trial and error here - I apologize for some mess from multiple runs.

If I cannot get rmatmat to work quickly here, I'll return to pre-rmammat version and try rmatmat in a different PR later.

lobpcg commented:
rmatmat addition actually seems to work, but opens a new can of worms with some errors from scipy/io/tests/test_fortran.py in https://github.com/scipy/scipy/blob/master/scipy/io/_fortran.py that I do not know how to fix. May be related fixes are
- https://github.com/yt-project/yt/pull/2152/files
- https://stackoverflow.com/questions/9644110/difference-between-parsing-a-text-file-in-r-and-rb-mode
Probably I should go back to to pre-rmammat version and try rmatmat in a different PR later.

pv commented:
rmatmat is unrelated to the fortranfile failures, that's some numpy dev version issue (also master branch is currently failing there).

lobpcg commented:
@pv -thanks! I am glad I have not started looking into these errors...
Then my rmatmat addition actually seems to work here.

pv commented on scipy/sparse/linalg/eigen/arpack/arpack.py:
```diff
@@ -1790,40 +1800,60 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
+        else:
+            X = np.random.RandomState(52).randn(min(A.shape), k)
+
+        eigvals, eigvec = lobpcg(XH_X, X, tol=tol ** 2, maxiter=maxiter,
```
Does lobpcg behave similarly as arpack with the default value `tol=0`? (ARPACK does `if tol == 0: tol = macheps`)

lobpcg replied:
Yes, the current lobpcg revision in master does 
```
residualTolerance = tol
...
if (residualTolerance is None) or (residualTolerance <= 0.0):
    residualTolerance = np.sqrt(1e-15) * n
```
so tol == 0 is treated sensibly

lobpcg commented:
@ilayn @mdhaber - could you please review? I would like to finish this while I can.
The Travis CI build failure is unrelated to this PR.

ilayn commented on scipy/sparse/linalg/eigen/arpack/arpack.py:
```diff
@@ -1790,40 +1800,60 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
-                for i, col in enumerate(V.T):
-                    out[i, :] = A.rmatvec(col.reshape(-1, 1)).T
-                return out.T
+                dtype = A.dot(np.zeros([m, 1])).dtype
```
Do we have to perform this full dot product here? If `m` is, say, 1e5 it is certainly a waste just to get the `dtype`. We can use a single element of `A` and 0. which is `float` for the result type. But if this is only to test complex or real we can directly test for the complex dtype.

pv replied:
If you reach here (object with no/None dtype attribute), what you have is a LinearOperator with unknown dtype. It's not possible to extract an element from it, except via matvec.

ilayn replied:
I agree but we are possibly upcasting to either `float` or `complex` in the end anyways because of `np.zeros` being `float`. so there is a binary choice. A less computation intensive option (asking specifically for dtype etc.) might serve better here.

pv replied:
The cost is one more matvec, whereas ARPACK/lobpcg will likely perform quite many of those.

What do you propose as an alternative? Raise error? That's going to be backward compatibility break, and I think the cost here (corner-case for users who did not explicitly specify dtype for their linear operator) is not large enough for that.

ilayn replied:
OK I can't judge how many matvecs are used in a typical go so if it is negligible then I am OK with it.

pv replied:
```
import numpy as np
from scipy.sparse.linalg import svds, LinearOperator
from scipy.sparse import diags

A = diags([np.arange(1, 10)**2], [0])
count = 0

def matvec(x):
    global count
    count += 1
    return A.dot(x)

def rmatvec(x):
    global count
    count += 1
    return A.T.dot(x)

op = LinearOperator(matvec=matvec, rmatvec=rmatvec, shape=A.shape)
res = svds(op, k=1, tol=1e-3, return_singular_vectors=False)
print("matvecs:", count)
```
prints 19.

ilayn replied:
Nice. Then my concern is not relevant indeed.

lobpcg replied:
@ilayn - thanks for the review! 
@pv thanks for resolving the issue!

lobpcg commented:
Added a comment in `svds` "The order of the singular values is not guaranteed." resolving https://github.com/scipy/scipy/issues/9643

mdhaber commented:
@lobpcg I saw your request, but unfortunately I cannot commit to providing further review of this PR, so please don't let it be held up by me. Thank you for your contribution!

lobpcg commented:
> @lobpcg I saw your request, but unfortunately I cannot commit to providing further review of this PR, so please don't let it be held up by me. Thank you for your contribution!

@mdhaber - you are still listed as a requested reviewer - is this a blocker? If so, can you remove it?

lobpcg commented:
@pv  - could you please have another look? 

In https://github.com/scipy/scipy/pull/3719/files you committed the "sophisticated detection of small eigenvalues" but somehow only for "LM", leaving "SM" branch unprotected from dividing by zero singular values. In this latest change, I just apply this part no matter if "LM" or "SM" .  This change, e.g., should supersede the stuck PR https://github.com/scipy/scipy/pull/8169/

lobpcg commented:
@mdhaber - you are still listed as a requested reviewer - is this a blocker? If so, can you remove it?

lobpcg commented:
Is there anything else I need ti do?

lobpcg commented:
Is there anything else I need ti do? Or ready to be merged?

rgommers commented:
This looks ready and it has been pretty well tested already. Since we're planning on branching the release in 4-5 days, let's get this in now. 

Thanks @lobpcg! And thanks all reviewers.

rgommers commented:
I've added a note to the release notes at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0:
_It is now possible to use the LOBPCG method in `sparse.linalg.svds`._

@lobpcg if you'd like to add some notes on other improvements, that would be much appreciated. The `rmatmat` one comes to mind, not sure off the top of my head what else (note that we don't note new features there, but not bug fixes unless there's something unusual that the users really need to know).

lobpcg commented:
> 
> 
> I've added a note to the release notes at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0:
> _It is now possible to use the LOBPCG method in `sparse.linalg.svds`._
> 
> @lobpcg if you'd like to add some notes on other improvements, that would be much appreciated. The `rmatmat` one comes to mind, not sure off the top of my head what else (note that we don't note new features there, but not bug fixes unless there's something unusual that the users really need to know).

@rgommers Done, added info on `rmatmat`  and float32 support in `lobpcg` in  he release notes at https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0: 

LinearOperator now supports new operation rmatmat for adjoint matrix-matrix multiplication, in addition to rmatvec in scipy.sparse.linalg.

Multiple stability updates enable float32 support in LOBPCG eigenvalue solver for symmetric and Hermitian eigenvalues problems in scipy.sparse.linalg.eigen.lobpcg.

rgommers commented:
Great, thanks @lobpcg
